central/delegatedregistryconfig/delegator/delegator.go

```diff
@@ -99,6 +109,29 @@ func (d *delegatorImpl) DelegateScanImage(ctx context.Context, imgName *storage.
 	return image, nil
 }
 
+// inferNamespace attempts to guess a namespace based on an image path, which is the convention used by images from
+// the OCP integrated registry. A namespace is returned only if it exists and the user has access. This inference
+// would be more accurate if done in the secured cluster however user access cannot be checked there. The namespace
+// is used by Sensor to pull additional secrets for authenticating to the image registry.
+func (d *delegatorImpl) inferNamespace(ctx context.Context, imgName *storage.ImageName, clusterID string) string {
+	// Extract namespace from image path following OCP integrated registry convention.
+	namespace := utils.ExtractOpenShiftProject(imgName)
```

Comment:
Is there a way to determine whether the given image name is _not_ from the OCP integrated registry (only way I currently see would be hardcoding but this is _meh_)?

My main point is I'd like to skip the search request below if it's not a OCP registry image. The namespace store can have quite a high volume and should avoid querying it if possible.

Reply:
That's what the comment was alluding to regarding performing the check in the secured cluster being more accurate. Within sensor we capture the 'hostname' of the OCP integrated registry and would be able to determine with certainty that the image is from that registry, but wouldn't be able to perform any access checks.  

Regarding hardcoding, although there is a default hostname (`image-registry.openshift-image-registry.svc:5000`), the name is configurable.  In at least one of the customer cases associated with this change the default hostname was changed - therefore I don't believe we can rely on the default.

> My main point is I'd like to skip the search request below if it's not a OCP registry image. The namespace store can have quite a high volume and should avoid querying it if possible.

Understood, can you think of an alternative? Don't believe there is a way around that unless we add comms to/from central/sensor to share what the current integrated registry hostname is - but even that has caveats as the registry can be referenced by various hostnames and IPs that all map to the same destination which may not be easily enumerated. Perhaps a benefit in this case is that delegation will only happen via ad-hoc requests. In the future I suspect that namespace could be explicitly 'specified' and not always inferred - at which point the namespace search / permission check should remain.

Does the namespace store degradation have a tracking ticket? Perhaps we can add an index for `name+clusterid` to help?

Reply:
When an image has `ClusterLocal` set, is it _always_ the internal registry or also other registries that are co-located and only accessible within the secured cluster? Maybe this could be used to distinguish whether we want to scan it or not.

From the code path however, I don't think we'll have the complete image proto there and only the name, so I don't think I'd have an alternative there (other than what you mentioned that Sensor will publish the registry information, but that seems a bit weird now that we've tried to do as much as possible within Sensor).

If there's no easy way to conditionally skip this, we can leave it in I suppose. The delegation shouldn't happen too often (hopefully) and we'll see whether queries will take a long time there.

For the namespace store: I don't think a ticket exists. One option to track the performance for this would be to create a metric for the function duration of the `DelegateScanImage` or even only the `inferNamespace` part ([ref here](https://github.com/stackrox/stackrox/blob/master/central/metrics/central.go#L280)). This way it's tracked / compared within our scale tests and if customers experience issues we can analyze the metrics within the diagnostic bundle.

Reply:
> When an image has ClusterLocal set, is it always the internal registry or also other registries that are co-located and only accessible within the secured cluster? Maybe this could be used to distinguish whether we want to scan it or not.

As of right now any image from any registry indexed via a secured cluster will have `ClusterLocal` set.

> From the code path however, I don't think we'll have the complete image proto there and only the name, so I don't think I'd have an alternative there (other than what you mentioned that Sensor will publish the registry information, but that seems a bit weird now that we've tried to do as much as possible within Sensor).

Correct, at the point making this decision will not have image metadata or scan results.

> One option to track the performance for this would be to create a metric for the function duration of the DelegateScanImage or even only the inferNamespace part ([ref here](https://github.com/stackrox/stackrox/blob/master/central/metrics/central.go#L280)). This way it's tracked / compared within our scale tests and if customers experience issues we can analyze the metrics within the diagnostic bundle.

Good idea, I created ROX-19678 for adding that metric, will cover in different PR